core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -895,6 +895,23 @@ public Optional<TableLayout> getNewTableLayout(Session session, String catalogNa
                 .map(layout -> new TableLayout(catalogHandle, transactionHandle, layout));
     }
 
+    @Override
+    public ConnectorTableMetadata getNewTableEffectiveMetadata(Session session, String catalogName, QualifiedObjectName table, List<ColumnMetadata> columns, Map<String, Object> properties, Optional<String> comment)
+    {
+        CatalogMetadata catalogMetadata = getRequiredCatalogMetadata(session, catalogName);
+        CatalogHandle catalogHandle = catalogMetadata.getCatalogHandle(session, table);
+        ConnectorMetadata metadata = catalogMetadata.getMetadataFor(session, catalogHandle);
+        ConnectorSession connectorSession = session.toConnectorSession(catalogHandle);
+        ConnectorTableMetadata newTableEffectiveMetadata = metadata.getNewTableEffectiveMetadata(
+                connectorSession,
+                table.asSchemaTableName(),
+                columns,
+                properties,
+                comment);
+        checkState(newTableEffectiveMetadata.getColumns().size() == columns.size(), "Table and query column count doesn't match");
```

Comment:
should we validate more things like column names etc? Maybe this method should only return a list of new types? Maybe this method should have following signature `Optional<Type> coerceNewTableColumn(Type type)` and it should be called per each column?